activesupport/lib/active_support/cache/memory_store.rb

```diff
@@ -121,30 +119,29 @@ def cached_size(key, entry)
         end
 
         def read_entry(key, **options)
-          entry = @data[key]
+          entry = nil
           synchronize do
```

Comment:
How about avoiding the lock on a miss:

```ruby
        def read_entry(key, **options)
          entry = @data[key]
          if entry
            synchronize do
              @data[key] = @data.delete(key)
            end

            entry = entry.dup
            entry.dup_value!
          end
          entry
        end
```

It seems to benchmark favourably:

### Before

```
Warming up --------------------------------------
           read:miss    42.720k i/100ms
          read:small    27.191k i/100ms
         write:small    18.488k i/100ms
Calculating -------------------------------------
           read:miss    429.763k (± 3.8%) i/s -      2.179M in   5.077735s
          read:small    274.424k (± 1.6%) i/s -      1.387M in   5.054514s
         write:small    184.886k (± 2.9%) i/s -    924.400k in   5.004337s
```

### After

```
Warming up --------------------------------------
           read:miss    47.788k i/100ms
          read:small    27.929k i/100ms
         write:small    19.005k i/100ms
Calculating -------------------------------------
           read:miss    485.357k (± 2.1%) i/s -      2.437M in   5.023825s
          read:small    275.148k (± 3.4%) i/s -      1.396M in   5.081857s
         write:small    189.401k (± 3.0%) i/s -    950.250k in   5.022170s
```

Reply:
I'm surprised it doesn't degrade the `read:small` performance, as it's now doing 2 lookups instead of one. But the hash lookups might be totally dwarfed by the rest of the operations.

Also the hash only has one key, maybe I should make it a bit bigger to have a more realistic benchmark.

But more importantly, the question is wether `dup` and `dup_entry` should be inside the mutex or not.

Reply:
Additionally, there's a potential race condition in your proposed code:

```ruby
        def read_entry(key, **options)
          entry = @data[key]
          # other thread delete that key
          if entry
            synchronize do
              # we now insert nil
              @data[key] = @data.delete(key)
            end
```

It's unlikely to happen with MRI, but could be a problem on other implementations.

Ultimately I don't think memory store is that much about performance, it's more of a development convenience, since the difference isn't that big, I would err on the side of caution.

Reply:
You're right that my version does slightly more work on a cache hit; I think `#[]` has a specialised instruction in MRI that's so well optimised that the difference is imperceptible in the benchmark.

Good catch on the race condition too! I'll admit I was primarily refactoring for style, but given that there's some potential issues, let's leave it as is. 👍